subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1065 @@
+	int ret;
+	int tmp[2];
+
+	ret = z_impl_zsock_socketpair(family, type, proto, tmp);
+	Z_OOPS(z_user_to_copy(sv, tmp, sizeof(tmp));
+	return ret;
+}
+
+#include <syscalls/zsock_spair_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
+/**
+ * Write data to one end of a @ref spair
+ *
+ * Data written on one file descriptor of a socketpair can be read at the
+ * other end using common POSIX calls such as read(2) or recv(2).
+ *
+ * If the underlying file descriptor has the @ref O_NONBLOCK flag set then
+ * this function will return immediately. If no data was written on a
+ * non-blocking file descriptor, then -1 will be returned and @ref errno will
+ * be set to @ref EAGAIN.
+ *
+ * Blocking write operations occur when the @ref O_NONBLOCK flag is @em not
+ * set and there is insufficient space in the @em remote @ref spair.pipe.
+ *
+ * Such a blocking write will suspend execution of the current thread until
+ * one of two possible results is received on the @em remote
+ * @ref spair.read_signal:
+ *
+ * 1) @ref SPAIR_SIG_DATA - data has been read from the @em remote
+ *    @ref spair.pipe. Thus, allowing more data to be written.
+ *
+ * 2) @ref SPAIR_SIG_CANCEL - the @em remote socketpair endpoint was closed
+ *    Receipt of this result is analagous to SIGPIPE from POSIX
+ *    ("Write on a pipe with no one to read it."). In this case, the function
+ *    will return -1 and set @ref errno to @ref EINTR.
+ *
+ * @param obj the address of an @ref spair object cast to `void *`
+ * @param buffer the buffer to write
+ * @param count the number of bytes to write from @p buffer
+ *
+ * @return on success, a number > 0 representing the number of bytes written
+ * @return -1 on error, with @ref errno set appropriately.
+ */
+static ssize_t spair_write(void *obj, const void *buffer, size_t count)
+{
+	ssize_t res;
+	bool have_local_sem = false;
+	bool have_remote_sem = false;
+
+	if (NULL == obj || NULL == buffer || 0 == count) {
+		LOG_DBG("invalid argument");
+		errno = EINVAL;
+		res = -1;
+		goto out;
+	}
+
+	struct spair *const spair = (struct spair *)obj;
+
+	res = k_sem_take(&spair->sem, K_FOREVER);
+	if (res < 0) {
+		LOG_ERR("failed to take local sem: %d", res);
+		errno = -res;
+		res = -1;
+		goto out;
+	}
+	have_local_sem = true;
+
+	struct spair *const remote = z_get_fd_obj(spair->remote,
+			(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	const bool is_connected = sock_is_connected(spair);
+	const size_t avail = is_connected ? spair_write_avail(spair) : 0;
+	const bool is_nonblock = sock_is_nonblock(spair);
+	const bool will_block = (count > avail) && !is_nonblock;
+
+	size_t bytes_written;
+
+	LOG_DBG("count: %u is_connected: %u avail: %u is_nonblock: %u "
+		"will_block: %u", count, is_connected, avail, is_nonblock,
+		will_block);
+
+	if (!is_connected) {
+		errno = EPIPE;
+		res = -1;
+		goto out;
+	}
+
+	res = k_sem_take(&remote->sem, K_FOREVER);
+	if (res < 0) {
+		errno = -res;
+		res = -1;
+		goto out;
+	}
+	have_remote_sem = true;
```

Comment:
empty line before

Reply:
> @jukkar @pfalcon @cfriedt I would really like to get support for socket pairs into 2.3. Any chance we can get there before Friday?

Yes - I'm making the last few changes requested now. Mainly just whitespace & style.